### ISSUE TITLE
File size for saved sim is too big

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#481]
After implementing the lightweight inventory, the size of a saved file has increased. 

One reason is that the object graph is different before and after the change.

Now we have a lot of lists that have object as elements, rather than having integers.

A brand new simulation has the saved file size of 680kB. 

```
00-Adir-01:013.566 (Config) Simulation : Saving the simulation as default.sim.
00-Adir-01:013.566 (Config) Simulation : heapSize: 384.00 MB    heapMaxSize: 1.50 GB    heapFreeSize: 251.58 MB
00-Adir-01:013.566 (Config) Simulation :            File size : 679.45 KB
00-Adir-01:013.566 (Config) Simulation : Done saving. The simulation resumes.
00-Adir-01:013.566 (Config) UIConfig : previous ui_settings.xml deleted.
00-Adir-01:013.566 (Config) UIConfig : Saving the ui_settings.xml.
```

After 5 sols, it needs 6.45MB.

```
00-Adir-05:322.757 (Config) Simulation : Saving the simulation as default.sim.
00-Adir-05:322.757 (Config) Simulation : heapSize: 660.00 MB    heapMaxSize: 1.50 GB    heapFreeSize: 269.78 MB
00-Adir-05:322.757 (Config) Simulation :            File size : 6.45 MB
00-Adir-05:322.757 (Config) Simulation : Done saving. The simulation resumes.
00-Adir-05:322.757 (Config) UIConfig : previous ui_settings.xml deleted.
00-Adir-05:322.757 (Config) UIConfig : Saving the ui_settings.xml.
```

May be it's about time we look into serialization.

**Comment by mokun** [comment]
So, my question is, how does UnitSet work to save space ?

**Comment by bevans2000** [comment]
It does save runtime memory only serialisation size. It only serialised the identifiers on Units instead of the object reference. That in itself doesn't save space but there is no need for the extra object mapping details that it uses to resolve the references when it is restored.

**Comment by bevans2000** [comment]
Of course one thing we need to be careful of is excessive object references in serialised objects. Whether an object is one of our classes or a base like String they still "expensive". Also objects that more instances like equipment produce better savings. This is why activities like #399 will also reduce the file size a bit because we will reuse the Locations. So instead of each Equipment having it's own copy of the Location they will all reference the same one as the Settlement.
I have noticed some overloaded properties as well where the same properties in the subclass as well as the base class.

**Comment by bevans2000** [comment]
The use of ```UnitSet``` should be more widespread to reduce the lookup table sizes and deserialization speed. However #649 should also be addressed first.

**Comment by mokun** [comment]
> The use of `UnitSet` should be more widespread to reduce the lookup table sizes and deserialization speed. 

But it's a set. Can you come up with a UnitList and UnitMap by any chance ?

**Comment by mokun** [comment]
I can see that inside `UnitSet`, somehow there's a List of INSTANCES.

```
/**
 * This class holds a set of Unit classes. It has a customised serialization 
 * format that only stores the Unit identifiers. The Unit references are restored
 * after deserialisation via the reinit method.
 * @param <T> Unit subclass.
 */
public class UnitSet<T extends Unit> 
	implements Set<T>, Serializable {

	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	 
    // The list of UnitSet that have been deserialised and need initialising
	@SuppressWarnings("rawtypes")
	private static final List<UnitSet> INSTANCES = new ArrayList<>();
```

**Comment by mokun** [comment]
> The use of UnitSet should be more widespread

UnitSet() has been used in EquipmentInventory, Rover's occupants, MarsSurface, Settlement's citizen/robots, vehicles, people.

![image](https://user-images.githubusercontent.com/1168584/181816693-a66fd6ef-0f1b-48d6-bf34-fe496fda9188.png)

**Comment by mokun** [comment]
I wish I can use it everywhere, but since `UnitSet` is a generic that extends Unit. I can't use it in a lot of places in which the object of a set is NOT an Unit. For instance, I can't use it for a set of `ScientificStudy` a person has.
```
/** The person's list of collaborative scientific studies. */
	private Set<ScientificStudy> collabStudies; 
```

Anyway to make it <T extend `something else` > ?

**Comment by mokun** [comment]
I guess we can only use it for Unit because of the following method `reload()`

```
    /**
     * Converts the temp list of Unit identifiers into the Unit references
     * into the internal set.
     * 
     * @param mgr
     */
	@SuppressWarnings("unchecked")
	private void reload(UnitManager mgr) {
		if (units == null) {
			units = new HashSet<>();
			
			for(int id : ids) {
				T found = (T) mgr.getUnitByID(id);
				if (found == null) {
					throw new IllegalStateException("Can not find Unit for id=" + id);
				}
				units.add(found);
			}
			ids = null;
		}
	}
```

But did you say that when starting a new sim, the id of an Alpha Base or any objects won't be the same ? 

I suppose only when reloading from a saved sim, the id of an unit is preserved and must be the same to be able to work.

**Comment by mokun** [comment]
> This implementation uses HashMap not the concurrency t version but I feel we over use the later anyway.

You said this earlier about `UnitSet` but I don't get what you meant. 

What I have wanted is to use `UnitSet` in UnitManager. However, most of the data structures in UnitManager are ConcurrentHashMap. 

I will convert ` Set<Unit> displayUnits;` to adopt `UnitSet` but that's all I can do.

But I suppose even within `UnitSet`, it relys on UnitManager's getUnitByID() to get back the Unit instance from its id.

So, can there be a version of `UnitSet` for map, such as `UnitMap` ?

```
		/** List of unit manager listeners. */
	private transient Set<UnitManagerListener> listeners;

	private transient ExecutorService executor;

	private transient Set<SettlementTask> settlementTasks = new HashSet<>();
	/** Map of equipment types and their numbers. */
	private Map<String, Integer> unitCounts = new HashMap<>();
	/** A map of all map display units (settlements and vehicles). */
	private Set<Unit> displayUnits;
	/** A map of settlements with its unit identifier. */
	private Map<Integer, Settlement> lookupSettlement;
	/** A map of sites with its unit identifier. */
	private Map<Integer, ConstructionSite> lookupSite;
	/** A map of persons with its unit identifier. */
	private Map<Integer, Person> lookupPerson;
	/** A map of robots with its unit identifier. */
	private Map<Integer, Robot> lookupRobot;
	/** A map of vehicle with its unit identifier. */
	private Map<Integer, Vehicle> lookupVehicle;
	/** A map of equipment (excluding robots and vehicles) with its unit identifier. */
	private Map<Integer, Equipment> lookupEquipment;
	/** A map of building with its unit identifier. */
	private Map<Integer, Building> lookupBuilding;


	/**
	 * Constructor.
	 */
	public UnitManager() {
		// Initialize unit collection
		lookupSite       = new ConcurrentHashMap<>();
		lookupSettlement = new ConcurrentHashMap<>();
		lookupPerson     = new ConcurrentHashMap<>();
		lookupRobot      = new ConcurrentHashMap<>();
		lookupEquipment  = new ConcurrentHashMap<>();
		lookupVehicle    = new ConcurrentHashMap<>();
		lookupBuilding   = new ConcurrentHashMap<>();

		listeners = new HashSet<>();
	}

```
.

**Comment by bevans2000** [comment]
We cant change ```UnutManager``` because each U it must have one hard link perserved otherwise it doesn't get serailsed.

Having a Map version could make sense but its only relevant for UnitManager but not needed as above. Then looking at List well do we have any real use cases were the order matters? Do we need the same Unit twice in a collection? I think it's no to both so every list should be converted to a Set

**Comment by mokun** [comment]
>  I think it's no to both so every list should be converted to a Set

Yes. I've converted a lot of HashSet to UnitSet that I found lately in BuildingManager.

Perhaps it's even better if we can find a way to comb through every class prior to serialization and encode/convert any object that extends Unit (except those in UnitManager) to an integer.

When deserializing, we'll decode it back to become an object again.

If we'll do it for the objects in UnitSet, we may as well do it for standalone objects if there's any merits to it.
